### PR TITLE
feat(model): Model change listener

### LINKED ISSUE #1168: feature: Detection of model change
The  model change detection is needed when:
* caching some derived information. The cache should be invalidated when model element which is source for the information in cache changes.
* model of higher abstraction level is based on spoon model. Here it needs to detect changes to be able to keep model up to date.

# Requirements
> R1 detects change of model element

For example if CtClass X changes name, then it should be possible to detect that on X

> R2 detects change of any child model element

For example if an statement is added/removed/modified in body of an method then it should be possible to detect that change on all parents: CtBlock, CtMethod, CtClass, CtPackage, CtModel.

> R3 detects change of model element referred by defined physical or computed relation

For example if method is added/removed/modified in CtClass A, then it should be possible to detect that on all classes, which directly or indirectly inherits/extends class A.

# Solutions
Let's collect here the ideas how to solve that.

## Solution 1 - Change events
We might fire event (call an method) on each change. The clients which needs to detect the change would listen for these events (using registered listener).

### Features
The clients is informed immediately when node is changed

### Problems
it can become performance problem, when it is used heavily and each change would fire many events ...
it can need a lot of memory to keep collections of listeners

## Solution 2 - CtElement#modelVersion attribute
We might have this code
```java
class CtElement {
   long modelVersion = 0;
  //returns true if this model or any child element was changed after time represented by modelVersionSnapshot
   boolean wasChangedAfter(long modelVersionSnapshot) {
      return modelVersionSnapshot<modelVersion;
  }
  public long getModelVersion() {
     return modelVersion;
  }
}
class CtModel {
  long version = 0;
  public long incrementVersion() {
    version++
    return version
  }
/*
* increments version of model and applies that version to modified element(s) and all parents
* @param e - the element whose content was changed
*/
  public long onElementChanged(CtElement... e) {
     long version = incrementVersion();
     e.modelVersion=version;
     e.forEachParent(parent->parent.modelVersion = version);
     return version;
  }
  public long getModelVersionSnapshot() {
     return version;
  }
}
```
### Features
The clients have to ask whether element was changed. The client is not informed immediately.

We have one model version counter. 
* It allows to apply new version to model elements using complex algorithms, which visits not only parents of changed element, but also related elements. Thanks to one version number, it is easy to detect that visiting algorithm already applied new version to the element and we can easily avoid infinite loops.
* It allows clients to store only one number - as snapshot - and detect if model was changed using that snapshot.
* It can be implemented thread safe - when another thread is modifying the model too, then higher version number is kept.

### Problems
Each model element has one long attribute.

### PR DESCRIPTION
**Pull request by tdurieux**
Close #1168

### CODE REVIEW

File: src/main/java/spoon/support/reflect/code/CtCommentImpl.java [hunk 1]
```diff
@@ -30,13 +33,21 @@
 	@MetamodelPropertyField(role = CtRole.COMMENT_TYPE)
 	private CommentType type;
 
+	public CtCommentImpl() {
+	}
+
+	protected CtCommentImpl(CommentType type) {
```
Comment (monperrus) [inline src/main/java/spoon/support/reflect/code/CtCommentImpl.java]: remove?

**Comment by monperrus** [comment]
rename `spoon.diff` in `spoon.experimental.modelobs`
put `ChangeFactory` in `spoon.experimental.modelobs`
add `spoon.experimental.package-info.java`:

    this is unstable code, subject to potential complete redesign, not only in snapshots but also in releases. you can use this code but you'll probably be impacted upon future changes. 

move addArgument to separate PR

**Comment by monperrus** [comment]
split ChangeFactory into ChangeFactory (empty on*) and `ListeningChangeFactory`

File: src/main/java/spoon/support/reflect/code/CtCommentImpl.java (hunk 1 above)
Comment (tdurieux) [inline src/main/java/spoon/support/reflect/code/CtCommentImpl.java]: I need it for CtJavaDocImpl

**Comment by tdurieux** [comment]
Done for me.

@pvojtechovsky Do you see something to improve?

**Comment by pvojtechovsky** [comment]
Nice code! I would not do it better. It is OK for me. Sometime I would prefer more comments ;-)

I like that it is by default disabled and that it adds no overhead if it is not needed.
I like Contexts and Actions and fast and low level FineModelChangeListener and higher level ActionBasedChangeListener.

Here are some topics for further discussion:
* Why FineModelChangeListener is not an interface?
* The `CtExecutable#setParameters(List<CtParameter<?>> parameters)` calls listener `deleteAll` and then `add` ... I can imagine some use cases where firing event about remove and add on the same element might be a problem (the element (e.g. parameter) will be marked dirty even if it was not really touched). What about another FineModelChangeListener method `onListSet`, which might internally call deleteAll and add, but implementor have a chance to detect what elements were really removed/added by listening on `onListSet`. Or we might use new UpdateMultipleAction, where value would be a kind of Collection.
* All the collections returned by any model getter should be readonly in future. So clients are forced to use appropriate modification methods.

The topics above might be discussed/solved later - after this big PR is merged.

Anyway thank You for this effort!

**Comment by tdurieux** [comment]
> I would prefer more comments ;-)

I know, I will try to add some

> Why FineModelChangeListener is not an interface?

Because it is really painful the create an interface when the things evolve a lot. We will create an interface when we will have some confidence in the design.

> The CtExecutable#setParameters(List<CtParameter<?>> parameters) calls listener deleteAll

I also don't like that and our replace uses these methods to make the change. The events are thus really awful. I made a clean method in the sniper PR to detect the delete all and the related add to keep only the relevant change.

> All the collections returned by any model getter should be readonly in future. So clients are forced to use appropriate modification methods.

That our final goal but it is really painful for the client.

**Comment by monperrus** [comment]
We're almost done :-)

- remove addArgument in interface/impl
- undo change in return type of `removeType`, `removeArgument`

File: src/main/java/spoon/reflect/factory/Factory.java [hunk 2]
```diff
@@ -156,6 +157,8 @@
 
 	QueryFactory Query();
 
+	FineModelChangeListener Change();
```
Comment (monperrus) [inline src/main/java/spoon/reflect/factory/Factory.java]: remove this change? The entry point is now clearly in Environment. (Thanks revapi)

**Comment by tdurieux** [comment]
Done for me

**Comment by monperrus** [comment]
Thanks @GerardPaligot for the intial work on this. Kudos @tdurieux for the heavyweight rebase and design improvements.

**Comment by pvojtechovsky** [comment]
Thanks @everybody for that great feature!

**Pull request merged**